services/context/repo.go

```diff
@@ -76,6 +76,58 @@ func CanWriteToBranch(ctx context.Context, user *user_model.User, repo *repo_mod
 	return issues_model.CanMaintainerWriteToBranch(ctx, permission, branch, user)
 }
 
+// RepositoryNotEditableMessage explains why a repository can not be edited by the user
+type RepositoryNotEditableMessage struct {
+	Reason     string
+	Repository string
+}
+
+// GetEditRepository returns the repository where the editor will actually write the
+// the edits to. This may be a fork of the repository owned by the user. If no repository
+// can be found for editing, either a reason or error is returned.
+func GetEditableRepository(ctx context.Context, user *user_model.User, repo *repo_model.Repository, branch string) (*repo_model.Repository, RepositoryNotEditableMessage, error) {
+	if CanWriteToBranch(ctx, user, repo, branch) {
+		return repo, RepositoryNotEditableMessage{}, nil
+	}
+
+	// If we can't write to the branch, try find a user fork to create a branch in instead
```

Comment:
We just need to check whether the user will be allowed to send pull request to the repository. The check could be done in the routers definition file. ref https://github.com/go-gitea/gitea/pull/34240/files#r2082805447

Reply:
I added this check now in `CanEnableEditor`, so that the same test can be used for both the router and templates. I also needed to check `CanWriteToBranch` instead of just can write code, so maintainers can edit files in forks.